compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -985,7 +985,11 @@ class Typer extends Namer
               case SAMType(sam)
               if !defn.isFunctionType(pt) && mt <:< sam =>
                 if (!isFullyDefined(pt, ForceDegree.all))
-                  ctx.error(ex"result type of closure is an underspecified SAM type $pt", tree.sourcePos)
+                  ctx.error(ex"result type of lambda is an underspecified SAM type $pt", tree.sourcePos)
+                else if (pt.classSymbol.is(FinalOrSealed)) {
```

Comment:
Should we move this check into the `SAMType` extractor?

Reply:
No I think it's better here. An extractor is for extracting, not reporting errors.

Reply:
I didn't mean to report an error in the extractor, just have the extractor return `None` on `sealed` and `final` classes. It seemed arbitrary to have a special error message for this. In `scalac` I get a type mismatch instead.:
```scala
scala> sealed trait Foo { def f(x: Int): Int }; val foo: Foo = (x: Int) => x
<console>:11: error: type mismatch;
 found   : Int => Int
 required: Foo
       sealed trait Foo { def f(x: Int): Int }; val foo: Foo = (x: Int) => x
```